x-pack/elastic-agent/pkg/composable/providers/docker/docker.go

```diff
@@ -0,0 +1,157 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package docker
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/common/bus"
+	"github.com/elastic/beats/v7/libbeat/common/docker"
+	"github.com/elastic/beats/v7/libbeat/common/safemapstr"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/errors"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/composable"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/config"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/core/logger"
+)
+
+func init() {
+	composable.Providers.AddDynamicProvider("docker", DynamicProviderBuilder)
+}
+
+type dockerContainerData struct {
+	container  *docker.Container
+	mapping    map[string]interface{}
+	processors []map[string]interface{}
+}
+type dynamicProvider struct {
+	logger *logger.Logger
+	config *Config
+}
+
+// Run runs the environment context provider.
+func (c *dynamicProvider) Run(comm composable.DynamicProviderComm) error {
+	watcher, err := docker.NewWatcher(c.logger, c.config.Host, c.config.TLS, false)
+	if err != nil {
+		// info only; return nil (do nothing)
+		c.logger.Infof("Docker provider skipped, unable to connect: %s", err)
+		return nil
+	}
+	startListener := watcher.ListenStart()
+	stopListener := watcher.ListenStop()
+	stoppers := map[string]*time.Timer{}
+	stopTrigger := make(chan *dockerContainerData)
+
+	if err := watcher.Start(); err != nil {
+		// info only; return nil (do nothing)
+		c.logger.Infof("Docker provider skipped, unable to connect: %s", err)
+		return nil
+	}
+
+	go func() {
+		for {
+			select {
+			case <-comm.Done():
+				startListener.Stop()
+				stopListener.Stop()
+
+				// Stop all timers before closing the channel
+				for _, stopper := range stoppers {
+					stopper.Stop()
+				}
+				close(stopTrigger)
+				return
+			case event := <-startListener.Events():
```

Comment:
Any particular reason of not preserving the previous approach of having helper methods?
```
case event := <-d.startListener.Events():
	d.startContainer(event)

case event := <-d.stopListener.Events():
	d.scheduleStopContainer(event)

case target := <-d.stopTrigger:
	d.stopContainer(target.container, target.metadata)
```

Reply:
Because I didn't want to set the channels on the instance of the controller and only wanted to do it in the function definition. Doing this way would have required me to pass all the variable instances into each one of the helper functions, and I didn't want to do that being that the helpers are very simple.